plugins/redfish/fu-ipmi-device.c

```diff
@@ -0,0 +1,573 @@
+					NULL, /* resp */
+					0,
+					NULL,
+					FU_IPMI_DEVICE_TIMEOUT,
+					error)) {
+		g_prefix_error(error, "failed to set user %02x name: ", user_id);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+gboolean
+fu_ipmi_device_set_user_enable(FuIpmiDevice *self, guint8 user_id, gboolean value, GError **error)
+{
+	guint8 op = value ? IPMI_PASSWORD_ENABLE_USER : IPMI_PASSWORD_DISABLE_USER;
+	const guint8 req[] = {user_id, op};
+
+	g_return_val_if_fail(FU_IS_IPMI_DEVICE(self), FALSE);
+	g_return_val_if_fail(user_id != 0x0, FALSE);
+
+	/* run transaction */
+	if (!fu_ipmi_device_transaction(self,
+					IPMI_NETFN_APP_REQUEST,
+					IPMI_SET_USER_PASSWORD,
+					req,
+					sizeof(req),
+					NULL, /* resp */
+					0,
+					NULL,
+					FU_IPMI_DEVICE_TIMEOUT,
+					error)) {
+		g_prefix_error(error, "failed to set user %02x enable: ", user_id);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+gboolean
+fu_ipmi_device_set_user_password(FuIpmiDevice *self,
+				 guint8 user_id,
+				 const gchar *password,
+				 GError **error)
+{
+	guint8 req[0x12] = {user_id, IPMI_PASSWORD_SET_PASSWORD};
+	gsize password_sz;
+
+	g_return_val_if_fail(FU_IS_IPMI_DEVICE(self), FALSE);
+	g_return_val_if_fail(user_id != 0x0, FALSE);
+	g_return_val_if_fail(password != NULL, FALSE);
+
+	/* copy into buffer */
+	password_sz = strlen(password);
+	if (!fu_memcpy_safe(req,
+			    sizeof(req),
+			    0x2, /* dst */
+			    (guint8 *)password,
+			    password_sz,
+			    0x0, /* src */
+			    password_sz,
+			    error)) {
+		g_prefix_error(error, "password invalid: ");
+		return FALSE;
+	}
+
+	/* run transaction */
+	if (!fu_ipmi_device_transaction(self,
+					IPMI_NETFN_APP_REQUEST,
+					IPMI_SET_USER_PASSWORD,
+					req,
+					sizeof(req),
+					NULL, /* resp */
+					0,
+					NULL,
+					FU_IPMI_DEVICE_TIMEOUT,
+					error)) {
+		g_prefix_error(error, "failed to set user %02x password: ", user_id);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+gboolean
+fu_ipmi_device_set_user_priv(FuIpmiDevice *self,
+			     guint8 user_id,
+			     guint8 priv_limit,
+			     guint8 channel,
+			     GError **error)
+{
+	const guint8 req[] = {channel, user_id, priv_limit, 0x0};
+
+	g_return_val_if_fail(FU_IS_IPMI_DEVICE(self), FALSE);
+	g_return_val_if_fail(user_id != 0x0, FALSE);
+	g_return_val_if_fail(channel <= 0x0F, FALSE);
+	g_return_val_if_fail(priv_limit <= 0x0F, FALSE);
+
+	/* run transaction */
+	if (!fu_ipmi_device_transaction(self,
+					IPMI_NETFN_APP_REQUEST,
+					IPMI_SET_USER_ACCESS,
+					req,
+					sizeof(req),
+					NULL, /* resp */
+					0,
+					NULL,
+					FU_IPMI_DEVICE_TIMEOUT,
+					error)) {
+		g_prefix_error(error,
+			       "failed to set user %02x privs of 0x%02x, 0x%02x: ",
+			       user_id,
+			       priv_limit,
+			       channel);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static void
+fu_ipmi_device_init(FuIpmiDevice *self)
+{
+	fu_device_set_physical_id(FU_DEVICE(self), "/dev/ipmi0");
```

Comment:
can this come from udev instead?  If not, do we need to also make sure that we load `impi-si` and `ipmi-devintf`?

Reply:
not really; there's no `udev_device` at all in FuUdevDevice, it's literally a hardcoded `device-file` we're using.

Reply:
> make sure that we load impi-si and ipmi-devintf

They seem loaded by default on Fedora and RHEL, is that a downstream thing?

Reply:
What is `CONFIG_IPMI_DEVICE_INTERFACE` set to?  I don't see anything for modaliases for the ipmi-devintf module in https://github.com/torvalds/linux/blob/5bfc75d92efd494db37f5c4c173d3639d4772966/drivers/char/ipmi/ipmi_devintf.c so unless it's compiled into the kernel I would think that module needs to get loaded _somehow_.

Reply:
> What is CONFIG_IPMI_DEVICE_INTERFACE set to

CONFIG_IPMI_DEVICE_INTERFACE=m

Maybe it's safer to manually load impi-si and ipmi-devintf? I'll do just that.

Reply:
Fixed.